src/commons/utils/InfiniteLoopReporter.ts

```diff
@@ -0,0 +1,162 @@
+import * as Sentry from '@sentry/browser';
+import { Context, NativeStorage, SourceError } from 'js-slang/dist/types';
+
+/**
+ * Parses the error and determines whether it is an infinite loop and
+ * if so, what kind of infinite loop.
+ *
+ * @returns {string} string containing the infinite loop classification.
+ * @returns {string} empty string, if the error is not an infinite loop.
+ */
+function infiniteLoopErrorType(errorString: string): string {
+  if (errorString.includes('Infinite recursion (or runtime error) detected')) {
+    if (errorString.includes('forget your base case')) return 'no_base_case';
+    else if (
+      errorString.includes('Did you call a value that is outside the range of your function?')
+    )
+      return 'input_out_of_domain';
+    else if (errorString.includes('Check your recursive function calls.')) return 'no_state_change';
+  }
+  // firefox
+  else if (errorString === 'InternalError: too much recursion') return 'stack_overflow';
+  // chrome
+  else if (errorString === 'RangeError: Maximum call stack size exceeded') return 'stack_overflow';
+  // edge
+  else if (errorString === 'Error: Out of stack space') return 'stack_overflow';
+  // error from transpiler
+  else if (errorString.includes('Potential infinite recursion detected'))
+    return 'source_protection_recursion';
+  else if (errorString.includes('Potential infinite loop detected'))
+    return 'source_protection_loop';
+  return '';
+}
+
+function getInfiniteLoopErrors(errors: SourceError[]) {
+  for (const error of errors) {
+    const errorType = infiniteLoopErrorType(error.explain());
+    if (errorType) return errorType;
+  }
+  return '';
+}
+
+/**
+ * stringifies an array, similar to JSON.stringify, but is able to
+ * 'stringify' functions in arrays (e.g. [(x=>x)]) -> "[(x=>x)]"
+ * JSON.stringify will return [null]). Also written iteratively as
+ * stringifying long (lispy) lists are likely to cause stack
+ * overflows. Has a depth limit in case of circular references.
+ *
+ */
+function stringifyArray(arr: any[]) {
+  let result = '';
+  const stack: [any[], integer][] = [[arr, 0]];
+  while (stack.length > 0) {
+    // arbitrary depth limit: 1000
+    if (stack.length > 1000) return result + ' ...REDACTED]';
+    const top = stack.pop() as [any[], integer];
+    const topArr = top[0];
+    let idx = top[1];
+    if (idx === 0) result += '[';
+    while (idx < topArr.length) {
+      const elem = topArr[idx];
+      if (Array.isArray(elem)) {
+        stack.push([topArr, idx + 1]);
+        stack.push([elem, 0]);
+        break;
+      } else if (typeof elem === 'function') {
+        result += elem.toString();
+      } else {
+        result += JSON.stringify(elem);
+      }
+      if (idx !== topArr.length - 1) result += ',';
+      idx++;
+    }
+    if (idx >= topArr.length) {
+      const top = stack[stack.length - 1];
+      result += ']';
+      if (top && top[1] !== top[0].length) result += ',';
+    }
+  }
+  return result;
+}
```

Comment:
You could probably use `JSON.stringify` directly, but with a replacer passed in as the 2nd argument that checks and removes circular structures and/or structures that are too deep.

Reply:
hmm my main concern was that JSON.stringify will quote the function code, and I couldn't find a satisfactory (non hacky) way to avoid this.
i.e. `[1,"two", (x)=>x]` will be stringified to `"[1, \"two\", \"(x)=>x\"]"` instead of the desired `"[1, \"two\", (x)=>x]"`